presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java

```diff
@@ -0,0 +1,350 @@
+    private static final String sourceColumnNameB = "source_col_b";
+    private static final ColumnHandle sourceColumnHandleB = new MockConnectorColumnHandle(sourceColumnNameB, INTEGER);
+
+    private static final SchemaTableName destinationTable = new SchemaTableName("target_schema", "target_table");
+    private static final String destinationColumnNameA = "destination_col_a";
+    private static final ColumnHandle destinationColumnHandleA = new MockConnectorColumnHandle(destinationColumnNameA, INTEGER);
+    private static final String destinationColumnNameB = "destination_col_b";
+    private static final ColumnHandle destinationColumnHandleB = new MockConnectorColumnHandle(destinationColumnNameB, INTEGER);
+
+    private static final Map<ColumnHandle, String> redirectionMappingA = ImmutableMap.of(sourceColumnHandleA, destinationColumnNameA);
+
+    private static final Map<ColumnHandle, String> redirectionMappingAB = ImmutableMap.of(
+            sourceColumnHandleA, destinationColumnNameA,
+            sourceColumnHandleB, destinationColumnNameB);
+
+    @Test
+    public void testRedirectionAfterProjectionPushdown()
+    {
+        // make the mock connector return a table scan on destination table only if
+        // the connector can detect that source_col_a is projected
+        try (LocalQueryRunner queryRunner = createLocalQueryRunner(
+                this::mockApplyRedirectAfterProjectionPushdown,
+                Optional.of(this::mockApplyProjection),
+                Optional.empty())) {
+            assertPlan(
+                    queryRunner,
+                    "SELECT source_col_a FROM test_table",
+                    output(
+                            ImmutableList.of("DEST_COL"),
+                            tableScan("target_table", ImmutableMap.of("DEST_COL", destinationColumnNameA))));
+
+            assertPlan(
+                    queryRunner,
+                    "SELECT source_col_a, source_col_b FROM test_table",
+                    output(
+                            ImmutableList.of("SOURCE_COLA", "SOURCE_COLB"),
+                            tableScan(TEST_TABLE, ImmutableMap.of("SOURCE_COLA", sourceColumnNameA, "SOURCE_COLB", sourceColumnNameB))));
+
+            assertPlan(
+                    queryRunner,
+                    "SELECT source_col_a FROM test_table WHERE source_col_a > 0",
+                    output(
+                            ImmutableList.of("DEST_COL"),
+                            filter(
+                                    "DEST_COL > 0",
+                                    tableScan(
+                                            equalTo(new MockConnectorTableHandle(destinationTable)),
+                                            TupleDomain.all(),
+                                            ImmutableMap.of("DEST_COL", equalTo(destinationColumnHandleA))))));
+        }
+    }
+
+    @Test
+    public void testRedirectionAfterPredicatePushdownIntoTableScan()
```

Comment:
could we add a case when filter is fully consumed and not rematerialized as part of redirection?